tests/e2e/test_parameterized_queries.py

```diff
@@ -0,0 +1,382 @@
+        ), "Compute no longer says it doesn't support native params. Remove conditional_protocol_patch from this file."
+
+    @patch(
+        "databricks.sql.client.Connection.server_parameterized_queries_enabled",
+        return_value=False,
+    )
+    def test_protocol_too_low(self, mock_parameterized_queries_enabled):
+        params = {"p": None}
+        with pytest.raises(
+            NotSupportedError,
+            match="Parameterized operations are not supported by this server. DBR 14.1 is required.",
+        ):
+            result = self._get_one_result(
+                ParameterApproach.NATIVE, params, bypass_patch=True
+            )
+
+    @both_approaches
+    def test_primitive_inferred_none(self, approach: ParameterApproach, inline_table):
+        params = {"p": None}
+        result = self._get_one_result(approach, params)
+        assert result.col == None
+
+    @both_approaches
+    def test_primitive_inferred_bool(self, approach: ParameterApproach, inline_table):
+        params = {"p": True}
+        result = self._get_one_result(approach, params)
+        assert result.col == True
+
+    @both_approaches
+    def test_primitive_inferred_integer(
+        self, approach: ParameterApproach, inline_table
+    ):
+        params = {"p": 1}
+        result = self._get_one_result(approach, params)
+        assert result.col == 1
+
+    def test_primitive_inferred_double(self):
+        params = {"p": 3.14}
+        result = self._get_one_result(ParameterApproach.NATIVE, params)
+        assert self._quantize(result.col) == self._quantize(3.14)
+
+    @both_approaches
+    def test_primitive_inferred_date(self, approach: ParameterApproach, inline_table):
+        # DATE in Databricks is mapped into a datetime.date object in Python
+        date_value = datetime.date(2023, 9, 6)
+        params = {"p": date_value}
+        result = self._get_one_result(approach, params)
+        assert result.col == date_value
+
+    @both_approaches
+    def test_primitive_inferred_timestamp(
+        self, approach: ParameterApproach, inline_table
+    ):
+        # TIMESTAMP in Databricks is mapped into a datetime.datetime object in Python
+        date_value = datetime.datetime(2023, 9, 6, 3, 14, 27, 843, tzinfo=pytz.UTC)
+        params = {"p": date_value}
+        result = self._get_one_result(approach, params)
+        assert result.col == date_value
+
+    @both_approaches
+    def test_primitive_inferred_string(self, approach: ParameterApproach, inline_table):
+        params = {"p": "Hello"}
+        result = self._get_one_result(approach, params)
+        assert result.col == "Hello"
+
+    @both_approaches
+    def test_primitive_inferred_decimal(
+        self, approach: ParameterApproach, inline_table
+    ):
+        params = {"p": Decimal("1234.56")}
+        result = self._get_one_result(approach, params)
+        assert result.col == Decimal("1234.56")
+
+    def test_dbsqlparam_inferred_none(self):
```

Comment:
`@both_approaches` is not used for any of the non-primitive test cases because INLINE mode is not compatible `DbsqlParameter` objects.